ports/stm/common-hal/sdioio/SDCard.c

```diff
@@ -0,0 +1,356 @@
+
+    const uint8_t sdio_clock_len = MP_ARRAY_SIZE(mcu_sdio_clock_list);
+    const uint8_t sdio_command_len = MP_ARRAY_SIZE(mcu_sdio_command_list);
+    const uint8_t sdio_data0_len = MP_ARRAY_SIZE(mcu_sdio_data0_list);
+    const uint8_t sdio_data1_len = MP_ARRAY_SIZE(mcu_sdio_data1_list);
+    const uint8_t sdio_data2_len = MP_ARRAY_SIZE(mcu_sdio_data2_list);
+    const uint8_t sdio_data3_len = MP_ARRAY_SIZE(mcu_sdio_data3_list);
+
+
+    // Loop over each possibility for clock.  Check whether all other pins can
+    // be used on the same peripheral
+    for (uint i = 0; i < sdio_clock_len; i++) {
+        const mcu_periph_obj_t *mcu_sdio_clock = &mcu_sdio_clock_list[i];
+        if (mcu_sdio_clock->pin != clock) {
+            continue;
+        }
+
+        int periph_index = mcu_sdio_clock->periph_index;
+
+        const mcu_periph_obj_t *mcu_sdio_command = NULL;
+        if (!(mcu_sdio_command = find_pin_function(mcu_sdio_command_list, sdio_command_len, command, periph_index))) {
+            continue;
+        }
+
+        const mcu_periph_obj_t *mcu_sdio_data0 = NULL;
+        if(!(mcu_sdio_data0 = find_pin_function(mcu_sdio_data0_list, sdio_data0_len, data[0], periph_index))) {
+            continue;
+        }
+
+        const mcu_periph_obj_t *mcu_sdio_data1 = NULL;
+        if(num_data > 1 && !(mcu_sdio_data1 = find_pin_function(mcu_sdio_data1_list, sdio_data1_len, data[1], periph_index))) {
+            continue;
+        }
+
+        const mcu_periph_obj_t *mcu_sdio_data2 = NULL;
+        if(num_data > 2 && !(mcu_sdio_data2 = find_pin_function(mcu_sdio_data2_list, sdio_data2_len, data[2], periph_index))) {
+            continue;
+        }
+
+        const mcu_periph_obj_t *mcu_sdio_data3 = NULL;
+        if(num_data > 3 && !(mcu_sdio_data3 = find_pin_function(mcu_sdio_data3_list, sdio_data3_len, data[3], periph_index))) {
+            continue;
+        }
+
+        if (reserved_sdio[periph_index-1]) {
+            sdio_taken = true;
+            continue;
+        }
+
+        self->clock = mcu_sdio_clock;
+        self->command = mcu_sdio_command;
+        self->data[0] = mcu_sdio_data0;
+        self->data[1] = mcu_sdio_data1;
+        self->data[2] = mcu_sdio_data2;
+        self->data[3] = mcu_sdio_data3;
+
+        return periph_index;
+    }
+
+    if (sdio_taken) {
+        mp_raise_ValueError(translate("Hardware busy, try alternative pins"));
+    } else {
+        mp_raise_ValueError_varg(translate("Invalid %q pin selection"), MP_QSTR_SDIO);
+    }
+}
+
+
+void common_hal_sdioio_sdcard_construct(sdioio_sdcard_obj_t *self,
+        const mcu_pin_obj_t * clock, const mcu_pin_obj_t * command,
+        uint8_t num_data, mcu_pin_obj_t ** data, uint32_t frequency) {
+
+    int periph_index = check_pins(self, clock, command, num_data, data);
+    SDIO_TypeDef * SDIOx = mcu_sdio_banks[periph_index - 1];
+
+    GPIO_InitTypeDef GPIO_InitStruct = {0};
+
+    // /* GPIOC and GPIOD Periph clock enable */
```

Comment:
remove if not used

Reply:
Currently we still turn all GPIO clocks on. I still need to add dynamic clock enables to claim-pin...

Reply:
removed